Developer3

ChervenIrina opened the pull request.

Commit:
Diplom3 commit

Commit:
Diplom3 commit

AlexKitIt commented on conftest.py:
```diff
@@ -0,0 +1,31 @@
+        driver = webdriver.Chrome()
+    elif request.param == "firefox":
+        options = Options()
+        options.binary = FirefoxBinary(r'E:\users\gorbunova_is\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\firefox.exe')
```
Нужно исправить: здесь и далее: абсолютного адресу до exeшника быть не должно, в CI это не заведется. Пути до бинарников описываются в PATH

ChervenIrina replied:
убрала абсолютные адреса, но firefox так и не получилось запустить (в PATH добален адрес до драйвера  ++ так же создала путь до exe самого браузера)

AlexKitIt commented on pages/base_page.py:
```diff
@@ -0,0 +1,50 @@
+    def wait_element(self, locator, time=20):
+        WebDriverWait(self.driver, time).until(EC.visibility_of_element_located(locator))
+
+    #def wait_element_visibility(self, locator, time=20):
```
Нужно исправить: закоммеченного кода быть не должно

ChervenIrina replied:
исправлено

AlexKitIt commented on pages/constructor_page.py:
```diff
@@ -0,0 +1,21 @@
+        self.wait_element(locator)
+        drag = self.driver.find_element(*locator)
+        drop = self.driver.find_element(*Locators.CONSTRUCTOR_BURGER_ORDER)
+        action_chains = ActionChains(self.driver)
+        action_chains.drag_and_drop(drag, drop).perform()
```
Нужно исправить: это необходимо вынести в отдельный метод в base. Не стоит использовать настолько низкоуровневые обращения в пейджах

ChervenIrina replied:
исправлено

AlexKitIt commented on pages/constructor_page.py:
```diff
@@ -0,0 +1,21 @@
+    @allure.step("Перетаскивание ингредиента в конструктор Бургера")
+    def add_ingredient_to_order(self, locator):
+        self.wait_element(locator)
+        drag = self.driver.find_element(*locator)
```
Нужно исправить: здесь и далее по проекту: все прямые обращения к driver необходимо вынести в base, там описать методы взаимодействия с ними и оттуда за счет наследования классов страниц от base переиспользовать
Кроме того, конкретно этот метод уже там описан

ChervenIrina replied:
исправлено

AlexKitIt commented on pages/forgot_and_reset_password.py:
```diff
@@ -0,0 +1,22 @@
+
+    @allure.step ('Открываем страницу Авторизации')
+    def go_to_site_login(self):
+        self.driver.get(Url.URL_LOGIN)
```
Нужно исправить: метод есть в base, стоит сделать урл принимаемым и тут его передавать. к driver обращаться из пейджей нельзя

ChervenIrina replied:
исправлено

AlexKitIt commented on pages/order_feed_page.py:
```diff
@@ -0,0 +1,20 @@
+            list_elements.append(element.text)
+        return list_elements
+
+    def close_form_order(self):
```
Нужно исправить: здесь и далее: над каждым методом пейджи ставится аннотация allure.step

ChervenIrina replied:
исправлено

AlexKitIt commented on tests/test_constructor_page.py:
```diff
@@ -0,0 +1,59 @@
+    def test_click_on_constructor_label(self, driver):
+        cp = ConstructorPage(driver)
+        cp.go_to_site()
+        cp.click_element(LPL.PERSONAL_ACCOUNT)
```
Нужно исправить: здесь и далее: методов с принимаемыми локаторами в тестах быть не должно, если это не вариант параметризации, они живут в методах пейджи. В тесте только создание объекта и вызов метода, без указания локаторов элементов с которым происходит взаимодействие. Тест должен быть максимально изолирован от логики выполнения

ChervenIrina replied:
исправлено

AlexKitIt commented on tests/test_forgot_password.py:
```diff
@@ -0,0 +1,46 @@
+        fp = ForgotPasswordPage(driver)
+        fp.go_to_site_login()
+        fp.click_element(FPL.LINK_FORGOT_PASSWORD)
+        assert driver.current_url == Url.URL_FORGOT_PASSWORD
```
Нужно исправить: здесь и далее: прямого обращения к driver, вызова его property, execute_script не нем и тд из тестов быть не должно. Для этого в base пишутся соответствующие методы и тут вызываются на объектах конкретных страниц

ChervenIrina replied:
исправлено

Commit:
Diplom3 fix commit:
1. conftest - убраны абсолютные пути к драйверу и exe firefox
2. в пейжах страниц сайта убрано явное обращение к driver (перенесено в base)
3. в тестах убраны явные обращения к локатарам (созданы методы в пейжах)
4. убран закомиченный код + над всеми методами проверено/добавлено allure.step

AlexKitIt commented on pages/constructor_page.py:
```diff
@@ -0,0 +1,59 @@
+import allure
+
+from constants import Url
+from locators.constructor_page_locators import Locators as CPL
+from locators.order_feed_page_locators import Locators as OFL
+from locators.login_page_locators import Locators as LPL
+from locators.base_page_locators import Locators as BPL
```
Нужно исправить: здесь и далее: так быть не должно. Это ConstructorPage, для него создан класс с элементами с этой страницы constructor_page_locators значит в этом классе могут быть только методы с использованием этих локаторов, так в этом и состоит вся суть реализации POM. Все методы класса должны быть ограничены только работой с элементами на конкретной пейдже.

Commit:
Diplom3 fix commit (two):
1. пересмотрено расположение локаторов, если локатор для нескольких пейжах, то перенесен в base_locators

AlexKitIt commented on locators/constructor_page_locators.py:
```diff
@@ -0,0 +1,14 @@
+    INGREDIENT_BUN = (By.XPATH, ".//a[@href = '/ingredient/61c0c5a71d1f82001bdaaa6d']")
+    INGREDIENT_COUNTER = (By.XPATH, "//a[@href='/ingredient/61c0c5a71d1f82001bdaaa6d']/div/p")
+    BUN_TOP = (By.XPATH, ".//div[@class='constructor-element constructor-element_pos_top']")
+    CONSTRUCTOR_BURGER_ORDER = (By.XPATH, '/html/body/div/div/main/section[2]/ul/li[2]')
```
Повторный коммент:
Нужно исправить: здесь и далее: использование порядкового номера элемента в dom является весьма нестабильным способом поиска, необходимо привязываться к атрибутам конкретного элемента
Кроме того ходить от html и использовать абсолютные пути нельзя - очень ненадежно, добавится на пути новый элемент - тест ляжет. Все подобные локаторы необходимо переписать

AlexKitIt commented on pages/constructor_page.py:
```diff
@@ -0,0 +1,58 @@
+import allure
+
+from constants import Url
+from locators.constructor_page_locators import Locators as CPL
+from locators.login_page_locators import Locators as LPL
+from locators.base_page_locators import Locators as BPL
```
Неиспользуемый импорт
